migration/migration.go

```diff
@@ -0,0 +1,175 @@
+package migration
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	logger "github.com/ipfs/go-log/v2"
+)
+
+var (
+	log = logger.Logger("migrations")
+
+	keyCurrentVersion = datastore.NewKey("version")
+)
+
+// Migrator ensures a datastore goes through all the needed
+// Migrations to upgrade its current version to the latest version.
+type Migrator struct {
+	ds         datastore.TxnDatastore
+	migrations map[int]Migration
+}
+
+// Migration runs a vA->v(A+1) migration.
+type Migration func(datastore.Txn) error
+
+// New returns a new Migrator.
+func New(ds datastore.TxnDatastore, migrations map[int]Migration) *Migrator {
+	m := &Migrator{
+		ds:         ds,
+		migrations: migrations,
+	}
+	return m
+}
+
+// Ensure detects the current datastore version, and runs all the known migrations
+// to upgrade to the latest known version.
+func (m *Migrator) Ensure() error {
+	currentVersion, emptyDS, err := m.getCurrentVersion()
+	if err != nil {
+		return fmt.Errorf("getting current version: %s", err)
+	}
+
+	targetVersion := m.getTargetVersion()
+
+	// If the database is empty, we can assume Powergate started fresh
+	// and requires no migration. Set current version to latest version
+	// in migrations script.
+	if emptyDS {
+		if err := m.bootstrapEmptyDatastore(targetVersion); err != nil {
+			return fmt.Errorf("bootstrapping empty database: %s", err)
+		}
+
+		return nil
+	}
+
+	log.Infof("Current datastore version is %d, target version %d", currentVersion, targetVersion)
+
+	if currentVersion == targetVersion {
+		return nil
+	}
+
+	for i := currentVersion + 1; i <= targetVersion; i++ {
+		log.Infof("Running %d migration...")
+		if err := m.run(i); err != nil {
+			return fmt.Errorf("running migration %d: %s", i, err)
+		}
+		log.Infof("Migration %d ran successfully")
+	}
```

Comment:
Each needed migration will run in its own transaction, so every version bumping should be safe.

Reply:
IIRC, MongoDB transactions are only available to replica sets. Just mentioning in case you've implemented `TxnDatastore` with `go-ds-mongo`.

Reply:
Yes, that's correct. But our MongoDB cluster is a replica set, right?

Reply:
Yep, it is. Not a problem, just mentioning. It would error and cleanly exit your migration if run on a single mongod

Reply:
Oh yeah, sure. In fact, in `go-ds-mongo` tests, the tests that involve Txn have to run in a local replica-set that I documented how to spinup since was kind of hairy. (And maybe is there some simpler solution), but yeah, a single `docker run mongo` will fail.

Reply:
Oh I see. I did a similar thing when starting the billing work, where I was relying on transactions (not any more). You can actually make a replica set with one member. I had a docker-compose file that does that, then runs the initiate command, making for a one liner. Maybe I can add that when doing the go-threads -> go-ds-mongo work.

Reply:
Cool, that sounds good to add to `go-ds-mongo` easier testing yep.